libraries/persistent_store/src/format.rs

```diff
@@ -0,0 +1,907 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// TODO(ia0): Remove when the module is used.
+#![allow(dead_code)]
+
+use crate::bitfield::*;
+use crate::{Storage, StorageIndex, StoreError, StoreResult};
+use alloc::vec::Vec;
+use core::cmp::min;
+use core::convert::TryFrom;
+
+type WORD = u32;
+
+/// Size of a word in bytes.
+const WORD_SIZE: usize = core::mem::size_of::<WORD>();
+
+/// Maximum size of a page in bytes.
+const MAX_PAGE_SIZE: usize = 4096;
+
+/// Maximum number of erase cycles.
+const MAX_CYCLE: usize = 65535;
+
+/// Maximum page index.
+///
+/// Thus the maximum number of pages in one more than this number.
+const MAX_PAGE: usize = 63;
+
+/// Maximum number of keys.
+const MAX_KEY: usize = 4095;
+
+/// Maximum length in bytes of a user payload.
+const MAX_VALUE_LEN: usize = 1023;
+
+/// Maximum number of updates per transaction.
+const MAX_UPDATES: usize = 31;
+
+/// Maximum number of words per virtual page.
+const MAX_VIRT_PAGE_SIZE: usize = div_ceil(MAX_PAGE_SIZE, WORD_SIZE) - CONTENT_WORD;
+
+/// Word with all bits set to one.
+const ERASED_WORD: u32 = 0xffffffff;
+
+/// Helpers for a given storage configuration.
+#[derive(Clone, Debug)]
+pub struct Format {
+    /// The size in bytes of a page in the storage.
+    page_size: usize,
+
+    /// The number of pages in the storage.
+    num_pages: usize,
+
+    /// The maximum number of times a page can be erased.
+    max_page_erases: usize,
+}
+
+impl Format {
+    /// Extracts the format from a storage.
+    ///
+    /// Returns `None` if the storage is not [supported].
+    ///
+    /// [supported]: struct.Format.html#method.is_storage_supported
+    pub fn new<S: Storage>(storage: &S) -> Option<Format> {
+        if Format::is_storage_supported(storage) {
+            Some(Format {
+                page_size: storage.page_size(),
+                num_pages: storage.num_pages(),
+                max_page_erases: storage.max_page_erases(),
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Returns whether a storage is supported.
+    ///
+    /// A storage is supported if the following conditions hold:
+    /// - The size of a word is 4 bytes.
+    /// - The size of a word evenly divides the size of a page.
+    /// - A page contains at least 8 words.
+    /// - A page contains at most [`MAX_PAGE_SIZE`] bytes.
+    /// - There is at least 3 pages.
+    /// - There is at most [`MAX_PAGE`]` + 1` pages.
+    /// - A word can be written at least twice between erase cycles.
+    /// - A page can be erased at most [`MAX_CYCLE`] times.
```

Comment:
It's maximum in both cases, i.e. smaller values are ok. For the number of erase cycles, we don't need a minimum number of erase cycles for the store to work. It works with any number including zero.

Reply:
Then let me clarify something: Will the store always start failing after `MAX_CYCLE` erase operations? What I think you want to say is:

> A page can be erased [`MAX_CYCLE`] times.

And you don't need to worry if you erased fewer than that? But it might still work if you erase more often?

Reply:
Oh I think I see what is confusing. I rephrased, please check if it's clearer. Maybe what is missing from the picture is that this function is about the storage configuration/specification, not the storage content. We are not checking whether an actual page has been erased some amount of times. We are checking that the flash specification exposes a maximum number of erase cycles that is at most `MAX_ERASE_CYCLE`.

Reply:
To me, the new description with "maximum" and "most" doesn't highlight how `MAX_ERASE_CYCLE` is about store specification and not actual hardware lifetime. So for inspiration, maybe more like:

> The maximum number of erase cycles (i.e. how often a page contributes to the total virtual capacity) is at most [`MAX_ERASE_CYCLE`].

Reply:
I see 2 aspects and I'm not sure what you suggest for each of them:
1. Clarifying "maximum number of erase cycles". This is what the "i.e." parenthesis is about. I actually think this is more confusing than anything, so I would suggest removing this parenthesis. If someone has read the `Storage` trait, they know what this refers to. And those conditions are about the storage, so this is the obvious place to get information.
2. Clarifying `MAX_ERASE_CYCLE`. There is currently no clarification about it. This is just a bound like the other ones above (`MAX_PAGE_INDEX`, etc.) so I don't see why this one would need particular explanation. This function is about checking that a storage is supported by the store, which should already clarify what those bounds are about.

I just removed the parenthesis for now to avoid possible confusion.

Reply:
Because those constant have documentation and are clickable, I wrote the clarification there. It's a bit of duplication but it's useful.